include/openssl/kdf.h

```diff
@@ -10,10 +10,94 @@
+
+size_t EVP_KDF_CTX_output_size(EVP_KDF_CTX *ctx);
+int EVP_KDF_CTX_derive(EVP_KDF_CTX *ctx, unsigned char *key, size_t *keylen);
+
+
+# define EVP_KDF_ALG_CTRL                       0x1100
+
+# define EVP_KDF_CTRL_PASS                      (EVP_KDF_ALG_CTRL)
+# define EVP_KDF_CTRL_SALT                      (EVP_KDF_ALG_CTRL + 1)
+# define EVP_KDF_CTRL_ITER                      (EVP_KDF_ALG_CTRL + 2)
+# define EVP_KDF_CTRL_MD                        (EVP_KDF_ALG_CTRL + 3)
+# define EVP_KDF_CTRL_KEY                       (EVP_KDF_ALG_CTRL + 4)
+# define EVP_KDF_CTRL_TLS_SECRET                (EVP_KDF_ALG_CTRL + 5)
+# define EVP_KDF_CTRL_TLS_SEED                  (EVP_KDF_ALG_CTRL + 6)
+# define EVP_KDF_CTRL_HKDF_INFO                 (EVP_KDF_ALG_CTRL + 7)
+# define EVP_KDF_CTRL_HKDF_MODE                 (EVP_KDF_ALG_CTRL + 8)
+# define EVP_KDF_CTRL_SCRYPT_N                  (EVP_KDF_ALG_CTRL + 9)
+# define EVP_KDF_CTRL_SCRYPT_R                  (EVP_KDF_ALG_CTRL + 10)
+# define EVP_KDF_CTRL_SCRYPT_P                  (EVP_KDF_ALG_CTRL + 11)
+# define EVP_KDF_CTRL_SCRYPT_MAXMEM_BYTES       (EVP_KDF_ALG_CTRL + 12)
+
+# define EVP_KDF_HKDF_MODE_EXTRACT_AND_EXPAND  0
+# define EVP_KDF_HKDF_MODE_EXTRACT_ONLY        1
+# define EVP_KDF_HKDF_MODE_EXPAND_ONLY         2
+
+# define EVP_KDF_CTX_set1_pass(kctx, pass, pwlen) \
+            EVP_KDF_CTX_ctrl(kctx, EVP_KDF_CTRL_PASS, pwlen, (void *)(pass))
+
+# define EVP_KDF_CTX_set1_salt(kctx, salt, saltlen) \
+            EVP_KDF_CTX_ctrl(kctx, EVP_KDF_CTRL_SALT, saltlen, (void *)(salt))
+
+# define EVP_KDF_CTX_set_iter(kctx, iter) \
+            EVP_KDF_CTX_ctrl(kctx, EVP_KDF_CTRL_ITER, iter, NULL)
+
+# define EVP_KDF_CTX_set_md(kctx, md) \
+            EVP_KDF_CTX_ctrl(kctx, EVP_KDF_CTRL_MD, 0, (void *)(md))
+
+# define EVP_KDF_CTX_set1_key(kctx, key, keylen) \
+            EVP_KDF_CTX_ctrl(kctx, EVP_KDF_CTRL_KEY, keylen, (void *)(key))
+
+# define EVP_KDF_CTX_set1_tls1_prf_secret(kctx, sec, seclen) \
+            EVP_KDF_CTX_ctrl(kctx, \
+                             EVP_KDF_CTRL_TLS_SECRET, seclen, (void *)(sec))
+
+# define EVP_KDF_CTX_add1_tls1_prf_seed(kctx, seed, seedlen) \
+            EVP_KDF_CTX_ctrl(kctx, \
+                             EVP_KDF_CTRL_TLS_SEED, seedlen, (void *)(seed))
+
+# define EVP_KDF_CTX_add1_hkdf_info(kctx, info, infolen) \
+            EVP_KDF_CTX_ctrl(kctx, \
+                             EVP_KDF_CTRL_HKDF_INFO, infolen, (void *)(info))
+
+# define EVP_KDF_CTX_set_hkdf_mode(kctx, mode) \
+            EVP_KDF_CTX_ctrl(kctx, EVP_KDF_CTRL_HKDF_MODE, mode, NULL)
+
+# define EVP_KDF_CTX_set_scrypt_N(kctx, n) \
+            EVP_KDF_CTX_ctrl_uint64(kctx, EVP_KDF_CTRL_SCRYPT_N, n)
+
+# define EVP_KDF_CTX_set_scrypt_r(kctx, r) \
+            EVP_KDF_CTX_ctrl_uint64(kctx, EVP_KDF_CTRL_SCRYPT_R, r)
+
+# define EVP_KDF_CTX_set_scrypt_p(kctx, p) \
+            EVP_KDF_CTX_ctrl_uint64(kctx, EVP_KDF_CTRL_SCRYPT_P, p)
+
+# define EVP_KDF_CTX_set_scrypt_maxmem_bytes(kctx, maxmem_bytes) \
```

Comment:
Should `EVP_KDF_CTX_set_scrypt_maxmem_bytes` be more generic to cater for future algorithms that also need a memory cap?
Rename to `EVP_KDF_CTX_set_maxmem_bytes`?

Reply:
Yes, I can imagine other algorithms needing a memory cap.

Reply:
I made that parameter generic.